### DISCUSSION TITLE
What is the purpose of vmq_ql ?

### DISCUSSION THREAD

**Discussion by mths1** [vernemq/vernemq#2090] [General]
I think it is all in the title. There is the vmq_ql plugin which seems to be largely undocumented, and I am curious what it does and what it is actualy used for...

**Comment by ioolkos** [comment]
Your finding the hidden gems :)
It's an "SQL like query interface for VerneMQ" (as you've probably seen in the app.src file app description). It's purpose is somewhat internal for now: serving the CLI interfaces that need to grab information about session state & more. An "SQL" query is not a query to an SQL database here, but is backed by live queries (to supervisors, for instance) to get that information. This is what `vmq_ql` allows.

`vmq_info`/`vmq_info_cli` are example users/implementations.

**Comment by ioolkos** [comment]
@mths1 there was even some work initially for more exposition (other than internal users).
For instance, you can start a REPL by hooking into Verne from another Erlang shell/node.

The VMQL Repl has an utterly weird format and behaviour, I wonder whether we could (or should) bring this to a more usable state. Could be fun :smile:


```
➜  erl -pa _build/default/lib/vmq_ql/ebin -noinput -remote_node VerneMQ@127.0.0.1  -s vmq_ql_repl start
Erlang/OTP 24 [erts-12.3.2.2] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit]
```
```
   _   ____  _______    __  
  | | / /  |/  / __ \  / /  
  | |/ / /|_/ / /_/ / / /__ 
  |___/_/  /_/\___\_\/____/ 
```
 ```                    
vmql> SELECT * FROM sessions WHERE client_id="testclient"
vmql> [#{'__mountpoint' => [],clean_session => true,client_id => <<"testclient">>,
                                                                                     deliver_mode => fanout,is_offline => false,is_online => true,
                                                                                                                                                     is_plugin => false,mountpoint => <<>>,node => <<"VerneMQ@127.0.0.1">>,
                                                            num_sessions => 1,offline_messages => 0,online_messages => 0,
                                                                                                                            queue_pid => <8383.628.0>,queue_started_at => 1677058392718,
                         session_pid => <8383.625.0>,session_started_at => 1677058392718,
                                                                                            statename => online}]
```

**Comment by codeadict** [comment]
I have been asking myself if the complexity of it is worth having this, it is a lot of code that could be replaced by an Erlang function call for its current internal usage. Having better observability and a Web dashboard can replace the need for a SQL-like language to query things

**Reply by ioolkos** [reply]
Since VMQL usage was always internal, I do not have much data whether people like to have an SQL-like interface or not. It might be an interesting differentiator (if made more useable)... or it might be some sort of vanity project :)
Which does not invalidate your points on internal and external use.

**Comment by ioolkos** [comment]
@codeadict @mths1 one thing that I'm currently exploring is to expose some of the management API functions over a `api/v2` REST module.
The VMQL layer certainly takes away performance, and I'm currently seeing users calling this in higher frequencies (for instance to query for the online status of a ClientId).
`api/v2` would maybe not have the full range of `api/v1` but be optimised for those REST calls that get used by scripts.

This can nicely use the enabling work on listeners, http_modules and api-key from @mths1. Thanks again for this! :)
